src/linker/Linker.Steps/LinkAttributesParser.cs

```diff
@@ -91,10 +92,17 @@ static string FormatCustomAttribute (CustomAttribute ca)
 			}
 		}
 
-		TypeDefinition? GenerateRemoveAttributeInstancesAttribute ()
+		TypeDefinition? GenerateRemoveAttributeInstancesAttribute (int ctorArgumentCount)
 		{
-			if (_context.MarkedKnownMembers.RemoveAttributeInstancesAttributeDefinition != null)
-				return _context.MarkedKnownMembers.RemoveAttributeInstancesAttributeDefinition;
+			TypeDefinition? td = null;
+
+			if (_context.MarkedKnownMembers.RemoveAttributeInstancesAttributeDefinition is TypeDefinition knownTypeDef) {
+				if (knownTypeDef.GetConstructorsOnType (_ => true).Any (m => m.Parameters.Count == ctorArgumentCount)) {
+					return knownTypeDef;
+				} else {
+					td = knownTypeDef;
```

Comment:
Here the type was recognized but a constructor with the number of parameters wasn't found, I think this is because you can delete attributes that match things like the first argument even if the constructor accepts 3 arguments. For example the test
```XML
<type fullname="Mono.Linker.Tests.Cases.LinkAttributes.TestConditionalRemoveAttribute">
	<attribute internal="RemoveAttributeInstances">
		<argument type="System.Object">
			<argument>remove1</argument>
		</argument>
	</attribute>
</type>
```
```C#
[TestConditionalRemove ("remove1", '1', 3)]  //This gets removed because the first arg matches
static void Removed_2 ()
{
}
```
The problem here is that I think there is nothing preventing from creating a malformed version of the type constructor, something that doesn't exist, for example:
```XML
<type fullname="Mono.Linker.Tests.Cases.LinkAttributes.TestConditionalRemoveAttribute">
	<attribute internal="RemoveAttributeInstances">
		<argument type="System.Object">
			<argument>arg1</argument>
		</argument>
		<argument type="System.Object">
			<argument>arg2</argument>
		</argument>
		<argument type="System.Object">
			<argument>arg3</argument>
		</argument>
		<argument type="System.Object">
			<argument>arg4</argument>
		</argument>
		<argument type="System.Object">
			<argument>arg5</argument>
		</argument>
	</attribute>
</type>

Reply:
Here, `knownTypeDef` is the type definition of `RemoveAttributeInstancesAttribute`, not the `TestConditionalRemoveAttribute` type. From my understanding, later in the linker it checks the arguments passed to the constructor of each instance of `TestConditionalRemoveAttribute` (the ones placed on the methods), against the arguments to the constructor of `RemoveAttributeInstancesAttribute` (which is placed on the `TestConditionalRemoveAttribute` definition by the xml). So we are generating a new ctor for that `RemoveAttributeInstancesAttribute` type to enable constructors of that to enable the xml to specify multiple arguments to match, not a ctor for `TestConditionalRemoveAttribute`.

So yes, it enables the xml with 5 arguments, but it's not generating a new ctor for the `TestConditionalRemoveAttribute` that shouldn't be there, and if the xml has more arguments than any of the `TestConditionalRemoveAttribute` ctors, then none are removed.

Reply:
For me, it's generating a new fake constructor with 5 parameters and adding it to _context.MarkedKnownMembers.RemoveAttributeInstancesAttributeDefinition.
There is no validation afterward, so it gets added and no warning is produced. Before it would warn saying that having more than 1 argument in the XML is not allowed, which is also not a great experience.
I do understand that TestConditionalRemoveAttribute won't have the constructor and therefore you cannot create an attribute with 5 constructors without getting a compiler error, but I think we should produce a warning for constructors that don't exist. Otherwise, we could have people thinking that what they wrote in the XML was correct and is actually doing nothing, or even worse they could write something that at some point is invalid but later on is deleting something that might not be needed to be removed.